packages/vscode-extension/src/debug/teamsAppInstallation.ts

```diff
@@ -0,0 +1,88 @@
+import { ExtensionErrors, ExtensionSource } from "../error";
+import * as StringResources from "../resources/Strings.json";
+
+import axios from "axios";
+import { returnSystemError, returnUserError } from "@microsoft/teamsfx-api";
+
+export async function showInstallAppInTeamsMessage(detected: boolean): Promise<boolean> {
+  const message = `${
+    detected
+      ? StringResources.vsc.localDebug.installApp.detection
+      : StringResources.vsc.localDebug.installApp.description
+  }${StringResources.vsc.localDebug.installApp.guide}`;
+  const result = await VS_CODE_UI.showMessage(
+    "info",
+    message,
+    true,
+    StringResources.vsc.localDebug.installApp.installInTeams,
+    StringResources.vsc.localDebug.installApp.continue
+  );
+  if (result.isOk()) {
+    if (result.value === StringResources.vsc.localDebug.installApp.installInTeams) {
+      const loginStatus = await graphLoginInstance.getStatus();
+      if (loginStatus.accountInfo?.upn === undefined) {
+        throw returnUserError(
+          new Error("M365 account info not found"),
+          ExtensionSource,
+          ExtensionErrors.GetTeamsAppInstallationFailed
+        );
+      }
+      const debugConfig = await commonUtils.getDebugConfig(true);
+      if (debugConfig?.appId === undefined) {
+        throw returnUserError(
+          new Error("Debug config not found"),
+          ExtensionSource,
+          ExtensionErrors.GetTeamsAppInstallationFailed
+        );
+      }
+      const url = `https://teams.microsoft.com/l/app/${debugConfig?.appId}?installAppPackage=true&webjoin=true&login_hint=${loginStatus.accountInfo.upn}`;
+      await VS_CODE_UI.openUrl(url);
+      return await showInstallAppInTeamsMessage(false);
+    } else if (result.value === StringResources.vsc.localDebug.installApp.continue) {
+      const internalId = await getTeamsAppInternalId();
+      return internalId === undefined ? await showInstallAppInTeamsMessage(true) : true;
+    } else if (result.value === StringResources.vsc.localDebug.installApp.cancel) {
+      return false;
+    }
+  }
+  return false;
+}
+
+export async function getTeamsAppInternalId(): Promise<string | undefined> {
+  const loginStatus = await graphLoginInstance.getStatus();
+  if (loginStatus.accountInfo?.upn === undefined || loginStatus.token === undefined) {
+    throw returnUserError(
+      new Error("M365 account info not found"),
+      ExtensionSource,
+      ExtensionErrors.GetTeamsAppInstallationFailed
+    );
+  }
+  const debugConfig = await commonUtils.getDebugConfig(true);
+  if (debugConfig?.appId === undefined) {
+    throw returnUserError(
+      new Error("Debug config not found"),
+      ExtensionSource,
+      ExtensionErrors.GetTeamsAppInstallationFailed
+    );
+  }
+  const url = `https://graph.microsoft.com/v1.0/users/${loginStatus.accountInfo.upn}/teamwork/installedApps?$expand=teamsApp,teamsAppDefinition&$filter=teamsApp/externalId eq '${debugConfig?.appId}'`;
```

Comment:
`upn` may contains special character. (See this [doc](https://docs.microsoft.com/en-us/graph/api/user-get?view=graph-rest-1.0&tabs=http#http-request)). Consider to use user id.

Reply:
Good point. `oid` in token is the user id. Can we assume every token contains `oid` field?

Reply:
Of course. And it's recommended to use `oid` instead of `upn`, by [doc](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens#validate-the-user-has-permission-to-access-this-data).

Reply:
Updated to use `oid`.